### PR TITLE
Add forward enzyme rules for init and solve

### PR DESCRIPTION
**Pull request by sharanry**
- Followup to #377 

TODO:

- [ ] Test against other algorithms 
- [ ] Error on unspported algorithms.

### CODE REVIEW

File: test/enzyme.jl [hunk 1]
```diff
@@ -161,4 +162,59 @@ Enzyme.autodiff(Reverse, f3, Duplicated(copy(A), dA), Duplicated(copy(b1), db1),
...
+@show fd_jac
+
+en_jac = map(onehot(b1)) do db1
+    eres = Enzyme.autodiff(Forward, fb, Duplicated(copy(b1), db1))
+    eres[1]
+end |> collect
+@show en_jac
+
+@test en_jac ≈ manual_jac
+@test en_jac ≈ fd_jac
```
Comment (sharanry) [inline test/enzyme.jl:196]: These tests fail due to numerical imprecision. Not entirely sure why.

File: test/enzyme.jl [hunk 2]
```diff
@@ -161,4 +162,59 @@ Enzyme.autodiff(Reverse, f3, Duplicated(copy(A), dA), Duplicated(copy(b1), db1),
...
+
+fd_jac = FiniteDiff.finite_difference_jacobian(fA, A) |> vec
+
+en_jac = map(onehot(A)) do dA
+    eres = Enzyme.autodiff(Forward, fA, Duplicated(copy(A), dA))
+    eres[1]
+end |> collect
+
+@test en_jac ≈ manual_jac
+@test en_jac ≈ fd_jac
```
Comment (sharanry) [inline test/enzyme.jl:220]: Same here

File: test/enzyme.jl [hunk 3]
```diff
@@ -161,4 +162,59 @@ Enzyme.autodiff(Reverse, f3, Duplicated(copy(A), dA), Duplicated(copy(b1), db1),
...
+
+fd_jac = FiniteDiff.finite_difference_jacobian(fA, A) |> vec
+
+en_jac = map(onehot(A)) do dA
+    eres = Enzyme.autodiff(Forward, fA, Duplicated(copy(A), dA))
+    eres[1]
+end |> collect
+
+@test_broken en_jac ≈ manual_jac
+@test_broken en_jac ≈ fd_jac
```
Comment (sharanry) [inline test/enzyme.jl:220]: These tests fail due to numerical imprecision. Not entirely sure why.

Comment (sharanry) [reply test/enzyme.jl:223]: df/db:
```
manual_jac = [-5.061916901336408, 1.5770609499033128, 5.446411839233853, 0.612648432464526]
fd_jac = [-5.061916947364807, 1.5770610570907593, 5.446411848068237, 0.6126484870910645]
en_jac = [-5.061916901336408, 1.5770609499033128, 5.446411839233853, 0.612648432464526]
```

df/dA:
```
manual_jac = [-1.0254607597283965, 33.43238752846228, 53.515223908678564, 28.42686093118159, 33.84013235610757, 22.569869144629088, 16.00129730157342, 24.207046306464335, 22.715575331910678, 26.035770464186847, 27.970857502536397, 25.5534616182086, 42.01248045607326, 20.02374061097621, 7.20819099154247, 23.21793954106552]
fd_jac = [-26.272539049386978, 8.185315012931824, 28.268145620822906, 3.1797885596752167, 8.593059450387955, -2.677203118801117, -9.245775252580643, -1.0400259792804718, -2.5314970314502716, 0.7886982262134552, 2.7237851321697235, 0.3063894212245941, 16.765405893325806, -5.22333163022995, -18.038884073495865, -2.029132694005966]
en_jac = [-26.272532992576195, 8.185315295614476, 28.268151675830794, 3.1797886983337977, 8.593060123259754, -2.6772030882187066, -9.24577493127438, -1.0400259263834648, -2.531496900937129, 0.7886982313390366, 2.7237852696885847, 0.30638938536080607, 16.76540822322545, -5.223331621871595, -18.038881241305333, -2.0291326917822765]
```

File: ext/LinearSolveEnzymeExt.jl [hunk 4]
```diff
@@ -9,6 +9,47 @@ using Enzyme
...
+
+    A = deepcopy(linsolve.val.A) #mutates after function is applied
+    res = func.val(linsolve.val; kwargs...)
+    
+    if RT <: Const
+        return res
+    end
+    
+    dres = deepcopy(res)
+    invA = inv(A)
```
Comment (ChrisRackauckas) [inline ext/LinearSolveEnzymeExt.jl:39]: never use `inv`

Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl]: addressed in d010911

File: ext/LinearSolveEnzymeExt.jl [hunk 5]
```diff
@@ -9,6 +9,47 @@ using Enzyme
...
+    db = linsolve.dval.b
+    dA = linsolve.dval.A
+    dres.u .= invA * (db - dA * res.u)
```
Comment (ChrisRackauckas) [inline ext/LinearSolveEnzymeExt.jl]: There's no need to refactorize here. This is just `A * u = db - dA * u`, or `u = A \ (db - dA * u)`. But this is just a `linsolve` call. Not only that, but it's the same operator as the one used in the forward pass, so you don't need to refactorize `A`. Therefore, this should simply use the same `linsolve` and do `linsolve.b = db - dA * u` and then `solve!`.

But I don't think the formula is correct. Isn't it just `linsolve.b = db` and then `solve!(linsolve)` then `du = linsolve.u`?

File: test/enzyme.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply test/enzyme.jl:223]: That's an expected difference due to summation order.

File: ext/LinearSolveEnzymeExt.jl (hunk 4 above)
Comment (ChrisRackauckas) [reply ext/LinearSolveEnzymeExt.jl]: Not fully, see https://github.com/SciML/LinearSolve.jl/pull/416#discussion_r1382633907. Your version has an extra factorization and is using the wrong linear solver.

File: ext/LinearSolveEnzymeExt.jl (hunk 5 above)
Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl]: But `linsolves.A` seem to mutate after the solve. That is why I had to add all this other stuff.

MWE:
```julia
using LinearSolve
A = rand(5,5)
b = rand(5)

prob = LinearProblem(A,b)
linsolve = init(prob)
@show linsolve.A
sol = solve!(linsolve)
@show linsolve.A
```
```
5×5 Matrix{Float64}:
 0.207897  0.0737386  0.220551   0.935437  0.883482
 0.362979  0.687049   0.220086   0.216771  0.145246
 0.538937  0.577133   0.965879   0.438704  0.689019
 0.348145  0.174776   0.0491639  0.145817  0.155529
 0.92258   0.637729   0.508441   0.166917  0.218296

5×5 Matrix{Float64}:
 0.92258    0.637729   0.508441   0.166917  0.218296
 0.393439   0.436142   0.0200455  0.1511    0.0593595
 0.584162   0.469106   0.659464   0.270316  0.533653
 0.225343  -0.160427   0.16558    0.877305  0.755451
 0.37736   -0.151045  -0.211799   0.185687  0.0548682
```

Comment (ChrisRackauckas) [reply ext/LinearSolveEnzymeExt.jl]: Yes, the mutation that it's doing is `lu!(A)`, i.e. its using the same memory in the representation of the LU-factorization. You want to let it use exactly the same mutated `A` via `cache.cacheval` in order to do the next solve, which is precisely what's done in the caching interface. See https://docs.sciml.ai/LinearSolve/stable/tutorials/caching_interface/

Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl]: Sounds good. And the simple  derivation of the formula is mentioned in https://scicomp.stackexchange.com/a/29421.

Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl]: Also, the correctness of the formula is checked with `FiniteDiff` in the tests - https://github.com/SciML/LinearSolve.jl/pull/416/files#diff-7c5a302dc0e55153407f2354959d201e28d2b9a61c4ef6bc86f05697a52b4cacR180-R196

File: test/enzyme.jl (hunk 3 above)
Comment (sharanry) [reply test/enzyme.jl]: addressed in 31745cc

File: ext/LinearSolveEnzymeExt.jl (hunk 4 above)
Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl]: addressed in 31745cc

File: ext/LinearSolveEnzymeExt.jl (hunk 5 above)
Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl]: addressed in 31745cc

**Review by ChrisRackauckas — APPROVED** [review]

File: ext/LinearSolveEnzymeExt.jl [hunk 6]
```diff
@@ -9,6 +9,50 @@ using Enzyme
 
 using EnzymeCore
 
+function EnzymeCore.EnzymeRules.forward(func::Const{typeof(LinearSolve.init)}, ::Type{RT}, prob::EnzymeCore.Annotation{LP}, alg::Const; kwargs...) where {RT, LP <: LinearSolve.LinearProblem}
+    @assert !(prob isa Const)
+    res = func.val(prob.val, alg.val; kwargs...)
```
Comment (wsmoses) [inline ext/LinearSolveEnzymeExt.jl:14]: Does init actually need a rule for forward mode?

File: ext/LinearSolveEnzymeExt.jl [hunk 7]
```diff
@@ -9,6 +9,50 @@ using Enzyme
...
+function EnzymeCore.EnzymeRules.forward(func::Const{typeof(LinearSolve.solve!)}, ::Type{RT}, linsolve::EnzymeCore.Annotation{LP}; kwargs...) where {RT, LP <: LinearSolve.LinearCache}
+    @assert !(linsolve isa Const)
+
+    res = func.val(linsolve.val; kwargs...)
+    
+    if RT <: Const
+        return res
+    end
+   
+    b = deepcopy(linsolve.val.b)
```
Comment (wsmoses) [inline ext/LinearSolveEnzymeExt.jl:37]: I don't think we need to copy here.

**Comment by ChrisRackauckas** [comment]
This looks like a real test failure.

File: ext/LinearSolveEnzymeExt.jl [hunk 8]
```diff
@@ -9,6 +9,52 @@ using Enzyme
...
+    if RT <: Const
+        return res
+    end
+    dres = func.val(prob.dval, alg.val; kwargs...)
+    dres.b .= res.b == dres.b ? zero(dres.b) : dres.b
+    dres.A .= res.A == dres.A ? zero(dres.A) : dres.A
+    if RT <: DuplicatedNoNeed
+        return dres
+    elseif RT <: Duplicated
+        return Duplicated(res, dres)
```
Comment (wsmoses) [inline ext/LinearSolveEnzymeExt.jl:24]: This doesn't handle batching atm

Comment (sharanry) [reply ext/LinearSolveEnzymeExt.jl:24]: tried to address this in 585cbb8. We can  make a followup PR for batch support.

**Comment by sharanry** [comment]
@ChrisRackauckas The test failures seem unrelated.

**Pull request merged**